net/xia/ppal_ether/main.c

```diff
@@ -18,12 +20,132 @@ static inline struct xip_ether_ctx *ctx_ether(struct xip_ppal_ctx *ctx)
 		: NULL;
 }
 
+/* ETHER-INTERFACE. */
+struct ether_interface {
+	struct net_device  *dev;
+	atomic_t           refcnt;
+	struct rcu_head    rcu_head;
+
+	/* Prevents racing conditions over list. */
+	spinlock_t         interface_lock;
+	struct list_head   list_interface_common_addr;
+};
+
+static inline struct ether_interface *__ether_get_rtnl(
+		const struct net_device *dev)
+{
+	return rtnl_dereference(dev->eth_ptr);
+}
+
+static inline struct ether_interface *ether_interface_get(
+		const struct net_device *dev)
+{
+	struct ether_interface *eint;
+
+	rcu_read_lock();
+	eint = rcu_dereference(dev->eth_ptr);
+	if (eint)
+		atomic_inc(&eint->refcnt);
+	rcu_read_unlock();
+
+	return eint;
+}
+
+static inline void ether_interface_hold(struct ether_interface *eint)
+{
+	atomic_inc(&eint->refcnt);
+}
+
+static inline void ether_interface_put(struct ether_interface *eint)
+{
+	if (atomic_dec_and_test(&eint->refcnt)) {
+		dev_put(eint->dev);
+		kfree(eint);
+	}
+}
+
 /* ETHER's virtual XID type. */
 static int ether_vxt __read_mostly = -1;
 
 /* ETHER_FIB table internal operations. */
 const struct xia_ppal_rt_iops *ether_rt_iops = &xia_ppal_list_rt_iops;
 
+/* Interface intialization and exit functions. */
+static struct ether_interface *eint_init(struct net_device *dev)
+{
+	struct ether_interface *eint;
+
+	ASSERT_RTNL();
+
+	eint = kzalloc(sizeof(*eint), GFP_KERNEL);
+	if (!eint)
+		return NULL;
+
+	eint->dev = dev;
+	dev_hold(dev);
+	atomic_set(&eint->refcnt, 0);
+
+	spin_lock_init(&eint->interface_lock);
+	INIT_LIST_HEAD(&eint->list_interface_common_addr);
+
+	ether_interface_hold(eint);
+	RCU_INIT_POINTER(dev->eth_ptr, eint);
+	return eint;
+}
+
+static void ether_interface_rcu_put(struct rcu_head *head)
+{
+	struct ether_interface *eint =
+			container_of(head, struct ether_interface, rcu_head);
+
+	ether_interface_put(eint);
+}
+
+static void eint_destroy(struct ether_interface *eint)
+{
+	ASSERT_RTNL();
+
+	RCU_INIT_POINTER(eint->dev->eth_ptr, NULL);
+	call_rcu(&eint->rcu_head, ether_interface_rcu_put);
```

Comment:
Thank to the comment on `ether_interface_put()`, you can just call `ether_interface_put()` instead of `call_rcu()`.